web/src/app/components/timeline/timeline.component.ts

```diff
@@ -0,0 +1,439 @@
+   * date will have a format of YYYY-MM-DD hh-mm, with single digits padded
+   * with a leading zero.
+   *
+   * @param d Holds one interval's data on the timeline.
+   */
+  private formatDate = (d: any) => {
+    const date = new Date(d);
+    const dateTimeFormat = new Intl.DateTimeFormat('en', {
+      year: 'numeric',
+      month: '2-digit',
+      day: '2-digit',
+      hour: '2-digit',
+      minute: '2-digit'
+    });
+
+    const formatToParts = 'formatToParts';
+    const [{ value: month }, ,
+           { value: day }, ,
+           { value: year }, ,
+           { value: hour }, ,
+           { value: minute }]
+      = dateTimeFormat[formatToParts](date);
+    
+    return `${year}-${month}-${day} ${hour}:${minute}`;
+  }
+
+  /**
+   * Composes content of the tooltip that will appear on hover.
+   *
+   * @param d Holds one interval's data on the timeline.
+   */
+  private getTooltipContent = (d: Item) => {
+    const duration = (d.endTime - d.startTime);
+
+    // Convert the duration, currently in milliseconds, to a human readable format
+    // with the largest unit in days and the smallest in seconds (e.g. passing in
+    // the value 361000 returns "6 minutes, 1 second")
+    const options = ({
+      round: true // Get rid of decimal places
+    } as HumanizeDurationOptions);
+    const output = TimelineComponent.HUMANIZER.humanize(duration, options);
+
+    // Return HTML representation of all required data.
+    return `<b>Push ID: ${d.pushID.slice(d.pushID.indexOf('@') + 1)}</b>
+      <br/>
+      <b>Final State: ${d.state}</b>
+      <br/>
+      <b>Start Time: ${this.formatDate(d.startTime)}</b>
+      <br/>
+      <b> End Time: ${this.formatDate(d.endTime)}</b>
+      <br/>
+      <b>Duration: ${output}</b>
+      `;
+  }
+
+  /**
+   * Create tooltip to display each interval's content.
+   *
+   * @param el Encasing element that holds the tooltip
+   */
+  private createTooltip = (el: any) => {
+    el.style('position', 'absolute')
+      .style('pointer-events', 'none')
+      .style('top', 0)
+      .style('opacity', 0)
+      .style('background', 'white')
+      .style('border-radius', '5px')
+      .style('box-shadow', '0 0 10px rgba(0,0,0,.25)')
+      .style('padding', '10px')
+      .style('line-height', '1.3')
+      .style('font', '11px sans-serif');
+  }
+
+  /**
+   * Creates a scrollable timeline with bars representing the duration of
+   * pushes. Every time this function is called, the previous timeline SVG is
+   * removed and a new one propagated to the screen. Timeline features include
+   * scrolling horizontally on the x-axis, zooming in and out along the x-axis,
+   * and a tooltip display when hovering over intervals. Each interval is
+   * color-coded according to its final state.
+   *
+   * The structure of the component is shown below:
+   * <div>
+   *   <svg>
+   *     <g>
+   *       <g>
+   *         <defs/>
+   *         <rect/>
+   *         <line/>
```

Comment:
Can you indicate (maybe using comments) what those tags are doing. [1] is one example where the name of the classes indicate what is going on.

[1] https://github.com/googleinterns/step189-2020/blob/66ca11d0dd319061f09625f7f1a4a801c0ff7717/web/src/app/components/cdf/cdf.component.ts#L90-L106